design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+`ksql.persistence.default.format.value` is set.
+
+Somewhat unrelated, but small, we propose also changing the `KAFKA_TOPIC` property from _required_ to 
+_optional_ in `CREATE STREAM` and `CREATE TABLE` statements. If not supplied, it will default to the 
+uppercase name of the stream or table.  This matches the behaviour of the property in 
+`CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements, where it is already optional.
+
+To additional configurations will be added to control topic names:
+
+* `ksql.persistence.topic.name.case`: if `lower`, defaults ksql to using lowercase topics names
+  where topic names are not explicitly provided. This includes matching a `CREATE TABLE` or 
+  `CREATE STREAM` statement to an existing topic. Default: `upper`.
+* `ksql.persistence.topic.name.case.insensitive`: if `true`, matching a `CREATE TABLE` or 
+  `CREATE STREAM` statement to an existing topic will be case insensitive. Multiple matches will 
+  result in an error. Default: `false`.
+
+With these changes there will no longer be any required `CREATE` properties. This means the `WITH` 
+clause for `CREATE` statements for streams and tables would be optional. We propose this makes the 
+syntax more intuitive for those already familiar with SQL. This would make a `CREATE TABLE` 
+statement ANSI standard.
+
+```sql
+SET 'ksql.persistence.default.format.key'='Avro';
+SET 'ksql.persistence.default.format.value'='Avro';
+
+CREATE TABLE USERS (
+  ID BIGINT PRIMARY KEY,
+  NAME STRING
+);
+
+CREATE STREAM BIDS (
+  ITEM_ID BIGINT KEY,
+  USER_ID BIGINT,
+  AMMOUNT INT
+);
+```
+
+### Implementation
+
+The system already serializes the key format of source, intermediate and sink topics as part of the
+query plan, meaning it should be fairly easily to plug in new formats. 
+
+Validation will be added to ensure only supported key formats are set, and that key column data types
+are supported by key formats.
+
+Most existing functionality should _just work_, as the key format only comes into play during 
+(de)serialization, (obviously). The only area where additional work is expected is joins.
+
+Joins require the binary key of both sides of the join to match and both sides to be delivered to 
+the same ksqlDB node.  The former normally ensuring the latter, unless a custom partitioning 
+strategy has been used.
+
+The introduction of additional key formats means that while the deserialized key from both sides of 
+a join may match, the serialized binary data may differ if the key serialization format is different.
+To accommodate this, ksqlDB will automagically repartition the right side of a join to match the key
+format of the left side.  Such repartitioning is possible and safe, even for tables, because the 
+logical key of the data will not have changed, only the serialization format. This ensures the 
+ordering of updates to a specific key are maintained across the repartition.
```

Comment:
It may still introduce out-of-order data across different keys and thus time-tracking ("stream-time"), and grace-period and retention-time might be affected. (But it might be the same issue if we pick the left-input for repartitioning?).

However, for a stream-table join, it might be simpler to repartition the stream-side? Most likely, the join won't happen on the original key and the stream must be repartioned anyway (and we can just change the format on-the-fly)?

Reply:
> It may still introduce out-of-order data across different keys and thus time-tracking ("stream-time"), and grace-period and retention-time might be affected. (But it might be the same issue if we pick the left-input for repartitioning?).

Sure, just like any other implicit repartitioning we do.  Not a lot we can do about it, other than call it out in the docs.

> However, for a stream-table join, it might be simpler to repartition the stream-side? Most likely, the join won't happen on the original key and the stream must be repartitioned anyway (and we can just change the format on-the-fly)?

Good point about leveraging any existing repartition to also change the format. (I'll add this to the KLIP) If there is no existing repartition, which side to choose is fairly arbitrary, however, I've outlined why I think it should be the right in my [comment](https://github.com/confluentinc/ksql/pull/6017#discussion_r473983198) about.

Reply:
i think the idea that, in a stream-table join, it's the stream that gets repartitioned (assuming any repartitioning needed at all of course) makes intuitive sense. If you think about it, you _should_ only join to the table on it's key anyway - unless you really understand the implications at least. 
Obviously table-table joins that require a re-partition should be delegated to the kstreams FK-join feature to handle. (I know that's not implemented into ksqlDB yet but it hopefully won't be too far in the future - seems sub-optimal to build some strange short-term workaround for it?)
That leaves stream-stream joins, for which I agree that some arbitrary rule is required, hopefully one that's easily remembered like "left side always gets repartitioned" :-)